File: AlgorithmsOnGraphs-UCSD/week2/toposort.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <stack>

using std::vector;
using std::pair;
using std::stack;

void dfs(vector<vector<int> > &adj, vector<int> &used, stack<int> &s, int x) {
    //write your code here
    used[x] = true;

    for(auto it = adj[x].begin(); it != adj[x].end(); it++){
        if(!used[*it]){
            dfs(adj, used, s, *it);
        }
    }
    s.push(x);
}     

vector<int> toposort(vector<vector<int> > adj) {
    vector<int> used(adj.size(), 0);
    vector<int> order;
    //write your code here
    stack<int> topological;
    
    for(int i = 0; i < adj.size(); i++){
        if(!used[i])
            dfs(adj, used, topological, i);
    }
    while(!topological.empty()){
        order.push_back(topological.top());
        topological.pop();
    }
    return order;
}

int main() {
    size_t n, m;
    std::cin >> n >> m;
    vector<vector<int> > adj(n, vector<int>());
    for (size_t i = 0; i < m; i++) {
        int x, y;
        std::cin >> x >> y;
        adj[x - 1].push_back(y - 1);
    }
    vector<int> order = toposort(adj);
    for (size_t i = 0; i < order.size(); i++) {
        std::cout << order[i] + 1 << " ";
    }
}

File: AlgorithmsOnGraphs-UCSD/week2/toposort.py
#Uses python3

import sys

def dfs(adj, used, order, x):
    #write your code here
    used[x] = True
    for node in adj[x]:
        if not used[node]:
            dfs(adj, used, order, node)
    order.append(x)
    
def toposort(adj):
    used = [0] * len(adj)
    order = []
    #write your code here
    for i in range(len(adj)):
        if not used[i]:
            dfs(adj, used, order, i)
    return order[::-1]

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    order = toposort(adj)
    for x in order:
        print(x + 1, end=' ')

